openedx/adg/lms/applications/management/commands/update_is_prerequisite_courses_passed.py

```diff
@@ -17,54 +18,145 @@ class Command(BaseCommand):
         Args:
-            pre_req_courses: List
+            prereq_course_groups: List of course groups
         Returns:
-            A querySet for users who should be checked for the update
+            A list of user ids who should be checked for the update
+        """
+        # Exclude users who passed courses or do not have application
+        users_to_be_checked_for_update = self.filter_user_by_application_hub_state()
+
+        for prereq_course_group in prereq_course_groups:
+            # Get user with valid course enrollment in prereq courses for current group
+            user_ids = self.get_users_with_active_course_enrollments(
+                users_to_be_checked_for_update, prereq_course_group
+            )
+
+            # Keep users who have active enrolments in each course group
+            users_to_be_checked_for_update = list(set(users_to_be_checked_for_update).intersection(user_ids))
+
+        return users_to_be_checked_for_update
+
+    def filter_user_by_application_hub_state(self):
+        """
+        Get users who does not have associated application hub and if they have, then they have not passed all prereq
+        courses
+        Returns:
+            A querySet for users who should be checked further
         """
         return User.objects.filter(
             Q(application_hub__isnull=True) | Q(application_hub__is_prerequisite_courses_passed=False),
-            courseenrollment__course__in=pre_req_courses,
+        ).values_list('id', flat=True)
+
+    def get_users_with_active_course_enrollments(self, users_to_be_checked_for_update, prereq_course_group):
+        """
+        Filter users from provided user ids. Filtered users that enrolled in one or more courses from course group
+        and has not un-enrollment course
+        Args:
+            users_to_be_checked_for_update: List of user ids
+            prereq_course_group: PrerequisiteCourseGroup model object
+        Returns:
+            A querySet for users who should be checked for the update
+        """
+        course_keys = prereq_course_group.course_keys()
+        return User.objects.filter(
+            id__in=users_to_be_checked_for_update,
+            courseenrollment__course__in=course_keys,
             courseenrollment__is_active=True,
-        ).annotate(
-            num_of_enrolled_pre_reqs=Count('username')
-        ).filter(
-            num_of_enrolled_pre_reqs=len(pre_req_courses)
-        )
+        ).values_list('id', flat=True).distinct()
+
+    def check_users_for_application_update(self, users_to_be_checked_for_update, prereq_course_groups):
+        """
+        Check check passed prereq courses for all filtered users
+        Args:
+            users_to_be_checked_for_update: List of user ids
+            prereq_course_groups: List of course groups
+        Returns:
+            None
+        """
+        for user_id in users_to_be_checked_for_update:
+            user = User.objects.get(id=user_id)
```

Comment:
We can avoid this query for all the users, We can query users before the loops and then use those objects in this loop.